### ISSUE TITLE
consistent-return and the "simple class instantiation" pattern

### ISSUE THREAD

**Issue by kevinoid** [eslint/eslint#5379]
Since b8b3b87 (first released in v2.0.0-alpha-1) introduced code path analysis (which is awesome, by the way!) the [Simple "Class" Instantiation pattern](http://ejohn.org/blog/simple-class-instantiation/) popularized by John Resig fails the `consistent-return` rule.  Consider his example, refactored into the more popular current formulation:

```js
function User(first, last){
  if (!(this instanceof User)) {
    return new User(first, last);
  }

  this.name = first + " " + last;
}
```

This now generates a `Expected to return a value at the end of this function (consistent-return)` error.  This is certainly correct, since the function has an inconsistent return.  Users can either disable the rule globally (a shame), disable it for every such constructor (a pain), `return this;` from every such constructor (uncommon, redundant, but correct), and/or limit the use of this pattern in future code (debatable).

Given the popularity of this pattern, the dubious value of the rule in this case, and the amount of code that will need to be changed to take advantage of `consistent-return`, I thought it might be worth considering adding an option to `consistent-return` to allow it.

What do you think?

**Comment by nzakas** [comment]
> the amount of code that will need to be changed to take advantage of consistent-return

What code are you talking about? Some context here would be useful.

Also, I'm not sure if such a check would be possible. @mysticatea what do you think?

**Comment by kevinoid** [comment]
> > the amount of code that will need to be changed to take advantage of consistent-return

> What code are you talking about? Some context here would be useful.

Any code which uses Simple "Class" Instantiation pattern

**Comment by kevinoid** [comment]
To get more at the gist of that question:  I've seen it in a lot of codebases.  I can start checking some of the more popular ones if you are concerned about how often that pattern is used.  I know it's frequently present in the Node runtime.

**Comment by kevinoid** [comment]
Re: technical feasibility.  My first thought at implementation would be to add an option which excludes `return new currentFunction()` from consistency checks.  It's not perfect, but I think it would have reasonably low false-positive/false-negative rates if enabled.

But it was just a thought.  I'm not tied to it.  Feel free to consider alternatives.

**Comment by mysticatea** [comment]
Thank you for this issue.

I got it, this is a well-known pattern in ES3/5.

Some rules (e.g. `no-invalid-this`) consider functions which start with an upper case character as a constructor. So I think this rule can behave in the same way.

- If a function name starts with an upper case character, this rule skips the check of the last code path.

It's feasible as adding the condition check to [this `if` statement](https://github.com/eslint/eslint/blob/master/lib/rules/consistent-return.js#L39-L43).

**Comment by kevinoid** [comment]
Thanks @mysticatea.  That works for me.  The `no-invalid-this` behavior sounds like a reasonable precedent and using the capitalization of the function name should be much simpler to implement than what I was thinking.
